File: src/main/java/icu/takeneko/omms/central/network/session/handler/builtin/controller/EndControllerConsoleRequestHandler.java
package icu.takeneko.omms.central.network.session.handler.builtin.controller;

import icu.takeneko.omms.central.network.session.request.Request;
import icu.takeneko.omms.central.network.session.response.Response;
import icu.takeneko.omms.central.permission.Permission;
import icu.takeneko.omms.central.network.session.SessionContext;
import icu.takeneko.omms.central.network.session.request.Request;
import icu.takeneko.omms.central.network.session.response.Response;
import icu.takeneko.omms.central.network.session.response.Result;
import icu.takeneko.omms.central.network.session.handler.builtin.BuiltinRequestHandler;
import icu.takeneko.omms.central.permission.Permission;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class EndControllerConsoleRequestHandler extends BuiltinRequestHandler {
    @Override
    public @Nullable Response handle(@NotNull Request request, @NotNull SessionContext session) {
        String id = request.getContent("consoleId");
        if (session.getControllerConsoleMap().containsKey(id)){
            var console = session.getControllerConsoleMap().get(id);
            console.close();
            session.getControllerConsoleMap().remove(id);
            return new Response().withResponseCode(Result.CONSOLE_STOPPED).withContentPair("consoleId",id);
        }else {
            return new Response().withResponseCode(Result.CONSOLE_NOT_EXIST).withContentPair("consoleId",id);
        }
    }

    @Override
    public @Nullable Permission requiresPermission() {
        return Permission.CONTROLLER_CONTROL;
    }
}

File: src/main/java/icu/takeneko/omms/central/network/session/handler/builtin/controller/EndControllerConsoleRequestHandler.kt
package icu.takeneko.omms.central.network.session.handler.builtin.controller

import icu.takeneko.omms.central.network.session.SessionContext
import icu.takeneko.omms.central.network.session.handler.builtin.BuiltinRequestHandler
import icu.takeneko.omms.central.network.session.request.Request
import icu.takeneko.omms.central.network.session.response.Response
import icu.takeneko.omms.central.permission.Permission

class EndControllerConsoleRequestHandler : BuiltinRequestHandler() {
    override fun handle(request: Request, session: SessionContext): Response {
        val id = request.getContent("consoleId")
        if (session.controllerConsoleMap.containsKey(id)) {
            val console = session.controllerConsoleMap[id]
            console!!.close()
            session.controllerConsoleMap.remove(id)
            return request.success().withContentPair("consoleId", id)
        } else {
            return request.fail().withContentPair("consoleId", id)
        }
    }

    override fun requiresPermission(): Permission {
        return Permission.CONTROLLER_CONTROL
    }
}
